subsys/bluetooth/mesh/scene_srv.c

```diff
@@ -0,0 +1,716 @@
+		bt_mesh_dtt_srv_transition_get(mod, &transition);
+	} else {
+		return;
+	}
+
+	if (tid_check_and_update(&srv->tid, tid, ctx)) {
+		BT_DBG("Duplicate TID");
+		scene_status_send(srv, ctx, BT_MESH_SCENE_SUCCESS);
+		return;
+	}
+
+	err = bt_mesh_scene_srv_set(srv, scene, &transition);
+	status = ((err == -ENOENT) ? BT_MESH_SCENE_NOT_FOUND :
+				     BT_MESH_SCENE_SUCCESS);
+
+	if (ack) {
+		scene_status_send(srv, ctx, status);
+	}
+
+	if (status == BT_MESH_SCENE_SUCCESS) {
+		/* Publish */
+		scene_status_send(srv, NULL, status);
+	}
+}
+
+static void handle_recall(struct bt_mesh_model *mod,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	scene_recall(mod, ctx, buf, true);
+}
+
+static void handle_recall_unack(struct bt_mesh_model *mod,
+				struct bt_mesh_msg_ctx *ctx,
+				struct net_buf_simple *buf)
+{
+	scene_recall(mod, ctx, buf, false);
+}
+
+static int scene_register_status_send(struct bt_mesh_scene_srv *srv,
+				      struct bt_mesh_msg_ctx *ctx,
+				      enum bt_mesh_scene_status status)
+{
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_SCENE_OP_REGISTER_STATUS,
+				 BT_MESH_SCENE_MSG_MINLEN_REGISTER_STATUS +
+					 2 * CONFIG_BT_MESH_SCENES_MAX);
+	bt_mesh_model_msg_init(&buf, BT_MESH_SCENE_OP_REGISTER_STATUS);
+	net_buf_simple_add_u8(&buf, status);
+	net_buf_simple_add_le16(&buf, srv->curr);
+
+	for (int i = 0; i < srv->count; i++) {
+		net_buf_simple_add_le16(&buf, srv->all[i]);
+	}
+
+	return model_send(srv->mod, ctx, &buf);
+}
+
+static void handle_register_get(struct bt_mesh_model *mod,
+				struct bt_mesh_msg_ctx *ctx,
+				struct net_buf_simple *buf)
+{
+	struct bt_mesh_scene_srv *srv = mod->user_data;
+
+	scene_register_status_send(srv, ctx, BT_MESH_SCENE_SUCCESS);
+}
+
+const struct bt_mesh_model_op _bt_mesh_scene_srv_op[] = {
+	{
+		BT_MESH_SCENE_OP_GET,
+		BT_MESH_SCENE_MSG_LEN_GET,
+		handle_get,
+	},
+	{
+		BT_MESH_SCENE_OP_RECALL,
+		BT_MESH_SCENE_MSG_MINLEN_RECALL,
+		handle_recall,
+	},
+	{
+		BT_MESH_SCENE_OP_RECALL_UNACK,
+		BT_MESH_SCENE_MSG_MINLEN_RECALL,
+		handle_recall_unack,
+	},
+	{
+		BT_MESH_SCENE_OP_REGISTER_GET,
+		BT_MESH_SCENE_MSG_LEN_REGISTER_GET,
+		handle_register_get,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+static uint16_t *scene_find(struct bt_mesh_scene_srv *srv, uint16_t scene)
+{
+	for (int i = 0; i < srv->count; i++) {
+		if (srv->all[i] == scene) {
+			return &srv->all[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void entry_recover(struct bt_mesh_scene_srv *srv, bool vnd,
+			  const struct scene_data *data)
+{
+	sys_slist_t *list = vnd ? &srv->vnd : &srv->sig;
+	struct bt_mesh_scene_entry *entry;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(list, entry, n) {
+		if (data->elem_idx != entry->mod->elem_idx) {
+			continue;
+		}
+
+		if (vnd) {
+			uint16_t company_id = sys_get_le16(data->data);
+
+			if (entry->mod->vnd.id != data->id) {
+				continue;
+			}
+
+			if (entry->mod->vnd.company != company_id) {
+				continue;
+			}
+
+			entry->type->recall(entry->mod, &data->data[2],
```

Comment:
It is probably better to write a comment what `2` means.

Reply:
Maybe even better to have a macro definition for that since `2` is used couple more times in the code.